pkg/vm/vm_core.go

```diff
@@ -1,15 +1,173 @@
 package vm
 
-import "github.com/lambdaclass/cairo-vm.go/pkg/vm/memory"
+import (
+	"errors"
+	"fmt"
+
+	"github.com/lambdaclass/cairo-vm.go/pkg/vm/memory"
+)
+
+type VirtualMachineError struct {
+	Msg string
+}
+
+func (e *VirtualMachineError) Error() string {
+	return fmt.Sprintf(e.Msg)
+}
 
 // VirtualMachine represents the Cairo VM.
 // Runs Cairo assembly and produces an execution trace.
 type VirtualMachine struct {
 	RunContext  RunContext
-	currentStep uint
+	CurrentStep uint
 	Segments    memory.MemorySegmentManager
 }
 
-func NewVirtualMachine() *VirtualMachine {
-	return &VirtualMachine{Segments: *memory.NewMemorySegmentManager()}
+func NewVirtualMachine() VirtualMachine {
+	return VirtualMachine{Segments: memory.NewMemorySegmentManager()}
+}
+
+type Operands struct {
+	Dst memory.MaybeRelocatable
+	Res *memory.MaybeRelocatable
+	Op0 memory.MaybeRelocatable
+	Op1 memory.MaybeRelocatable
+}
+
+type OperandsAddresses struct {
+	Dst_addr memory.Relocatable
+	Op0_addr memory.Relocatable
+	Op1_addr memory.Relocatable
+}
+
+// ------------------------
+//  Deduced Operands funcs
+// ------------------------
+
+type DeducedOperands struct {
+	Operands uint8
+}
+
+func (vm *VirtualMachine) OpcodeAssertions(instruction Instruction, operands Operands) error {
+	switch instruction.Opcode {
+	case AssertEq:
+		if operands.Res == nil {
+			return &VirtualMachineError{"UnconstrainedResAssertEq"}
+		}
+		if !operands.Res.IsEqual(&operands.Dst) {
+			return &VirtualMachineError{"DiffAssertValues"}
+		}
+	case Call:
+		new_rel, err := vm.RunContext.Pc.AddUint(instruction.size())
+		if err != nil {
+			return err
+		}
+		returnPC := memory.NewMaybeRelocatableRelocatable(new_rel)
+
+		if !operands.Op0.IsEqual(returnPC) {
+			return &VirtualMachineError{"CantWriteReturnPc"}
+		}
+
+		returnFP := vm.RunContext.Fp
+		dstRelocatable, _ := operands.Dst.GetRelocatable()
+		if !returnFP.IsEqual(&dstRelocatable) {
+			return &VirtualMachineError{"CantWriteReturnFp"}
+		}
+	}
+
+	return nil
+}
+
+func (deduced *DeducedOperands) set_dst(value uint8) {
+	deduced.Operands = deduced.Operands | value
+}
+
+func (deduced *DeducedOperands) set_op0(value uint8) {
+	deduced.Operands = deduced.Operands | value<<1
+}
+
+func (deduced *DeducedOperands) set_op1(value uint8) {
+	deduced.Operands = deduced.Operands | value<<2
+}
+
+//------------------------
+//  virtual machines funcs
+// ------------------------
+
+func VmNew(run_context RunContext, current_step uint, segments_manager memory.MemorySegmentManager) VirtualMachine {
+	return VirtualMachine{
+		RunContext:  run_context,
+		CurrentStep: current_step,
+		Segments:    segments_manager,
+	}
+}
```

Comment:
This function is redundant, we already have `NewVirtualMachine`
If this is used to build tests then we should move it to the vm_test package

Reply:
that function does not allow me to pass a run context to the virtual machine and i need it in order to do the testing, maybe i can move it to that package but i required it.